sail/src/main/java/org/apache/rya/rdftriplestore/inference/HasValueVisitor.java

```diff
@@ -0,0 +1,96 @@
+import org.apache.rya.rdftriplestore.utils.FixedStatementPattern;
+import org.openrdf.model.Resource;
+import org.openrdf.model.URI;
+import org.openrdf.model.Value;
+import org.openrdf.model.vocabulary.OWL;
+import org.openrdf.model.vocabulary.RDF;
+import org.openrdf.query.algebra.StatementPattern;
+import org.openrdf.query.algebra.TupleExpr;
+import org.openrdf.query.algebra.Var;
+
+public class HasValueVisitor extends AbstractInferVisitor {
+    public HasValueVisitor(RdfCloudTripleStoreConfiguration conf, InferenceEngine inferenceEngine) {
+        super(conf, inferenceEngine);
+        include = true;
+    }
+
+    @Override
+    protected void meetSP(StatementPattern node) throws Exception {
+        final Var subjVar = node.getSubjectVar();
+        final Var predVar = node.getPredicateVar();
+        final Var objVar = node.getObjectVar();
+        // We can reason over two types of statement patterns:
+        // { ?var rdf:type :Restriction } and { ?var :property ?value }
+        // Both require defined predicate
+        if (predVar != null && predVar.getValue() != null) {
+            final URI predURI = (URI) predVar.getValue();
+            if (RDF.TYPE.equals(predURI) && objVar != null && objVar.getValue() != null
+                    && objVar.getValue() instanceof Resource) {
+                // If the predicate is rdf:type and the type is specified, check whether it can be
+                // inferred using any hasValue restriction(s)
+                final Resource objType = (Resource) objVar.getValue();
+                final Map<URI, Set<Value>> sufficientValues = inferenceEngine.getHasValueByType(objType);
+                if (sufficientValues.size() > 0) {
+                    final Var valueVar = new Var("v-" + UUID.randomUUID());
+                    TupleExpr currentNode = node.clone();
+                    for (URI property : sufficientValues.keySet()) {
+                        final Var propVar = new Var(property.toString(), property);
+                        final TupleExpr valueSP = new DoNotExpandSP(subjVar, propVar, valueVar);
+                        final FixedStatementPattern relevantValues = new FixedStatementPattern(objVar, propVar, valueVar);
+                        for (Value value : sufficientValues.get(property)) {
+                            relevantValues.statements.add(new NullableStatementImpl(objType, property, value));
+                        }
+                        currentNode = new InferUnion(currentNode, new InferJoin(relevantValues, valueSP));
+                    }
+                    node.replaceWith(currentNode);
+                }
+            }
+            else {
+                // If the predicate has some hasValue restriction associated with it, then finding
+                // that the object belongs to the appropriate type implies a value.
+                final Map<Resource, Set<Value>> impliedValues = inferenceEngine.getHasValueByProperty(predURI);
```

Comment:
All classes with such a restriction, and the values associated with them, and we're trying to match objVar to the value. At this point, the check RDF.TYPE.equals(predURI) failed, which means there isn't any reason to expect objVar to be a class. Instead, we're looking at things like "?subj :p1 ?obj" and saying: If there is a property restriction on p1, involving class C1 and value v1, then this expresses the fact that all members x of :C1 implicitly have the triple (x, :p1, :v1). Therefore, for any x found to belong to :C1, {?subj=x, ?obj=:v1} is a correct solution to the original query. This map from Resources to Values would include an entry mapping C1 to the set {v1}.